### ISSUE TITLE
Queued Action event cooldown check failing for Priest

### ISSUE THREAD

**Issue by scamille** [simulationcraft/simc#5365]
### Describe the bug
Priest simulation fails assert at https://github.com/simulationcraft/simc/blob/97b7dccaa6f6413670fd51debf19bc7eb48a7ad0/engine/action/sc_action.cpp#L93
```simc: /simc/engine/action/sc_action.cpp:93: virtual void {anonymous}::queued_action_execute_event_t::execute(): Assertion `action->cooldown->ready <= sim().current_time()' failed.```
Crashing action is mind_blast

### To Reproduce
```./simc iterations=100000 T25_Priest_Shadow.simc fight_style=HeavyMovement output=/dev/null cleanup_threads=1 talents=0000300 default_actions=1```

Does not seem to be reproducible with a specific seed, just needs a lot (10'000+) iterations to occur.

### Additional information
Shadowlands 97b7dccaa6f6413670fd51debf19bc7eb48a7ad0

Originally reported at WarcraftPriests/sl-shadow-priest#104

Might be related to 3287f9f9a78895b509e61a0d6874f25aaf5d96b3

With dark thoughts, the cooldown object of Mind Blast is replaced temporarily during its own execution with dark thoughts buff up. That might mess with it and cause the problem.

-> Not yet sure what is happening here, mainly wanted to document how to reproduce it first.

cc @seanpeters86 @navv1234 @vituscze

**Comment by vituscze** [comment]
Managed to reproduce this with some debug output and the problem seems fairly clear. If Voidform expiration eats a charge, and Mind Blast was already queued, it needs to stops the queued execute.

```
19.305 Executing event: buff_expiration(#111)
19.305 Player 'T25_Priest_Shadow' Buff voidform ticks (15 of 15).
19.305 Player 'T25_Priest_Shadow' has 1.0 charges of mind blast as voidform ended
19.305 T25_Priest_Shadow next cast while casting cooldown ready at -3600.25
19.305 T25_Priest_Shadow next cast while casting cooldown ready at 22.782
19.305 Player 'T25_Priest_Shadow' has 0.0 charges of mind blast after voidform ended
19.305 Player 'T25_Priest_Shadow' loses voidform
19.305 Canceled event: buff_tick(#511)
19.305 Executing event: Queued-Action-Execute(#533)
```

**Comment by seanpeters86** [comment]
@vituscze actually strangely enough in-game it lets you finish the cast in that case 👀, so I guess we need to let that through? Maybe converting to just using charges makes more sense here

**Comment by vituscze** [comment]
No, queued execute isn't cast finish, it's cast start

**Comment by scamille** [comment]
I had a discussion yesterday with Publik and it might be the most sensible to just change the dark thoughts buff back to changing mind blast cooldown charges, without actually resetting the whole cooldown. That would also better match the in-game implementation.

As far as I understand the original reason to not do it that way was just because the SimC cd charges system did not handle dynamically adjusting them correctly. But we should just fix that.

I didn't really question the implementation in 3287f9f much at the time, but I now think it should be changed.

**Comment by Saeldur** [comment]
I originally avoided using the charge system with dark thoughts to try and avoid errors like this crash here. However, it might be worth looking into switching it all to be a charge system. I tried an implementation last week but it wasn't working properly and was producing different than expected results. Another thing to be aware of is that constantly adjusting the mindblast cooldown/charges is possibly expensive computationally.

**Comment by scamille** [comment]
@seanpeters86 did you already change the implementation to a charge based system? I haven't been tracking changes lately.

**Comment by vituscze** [comment]
Should be fixed by 15b9334a7aeafa2b61915aa59d0516e2d988cbcc
